tikv/tikv
# external_storage: move cloud code to separate cloud crate

TODO:
- [X] merge [CloudDynamic kvproto upstream](https://github.com/pingcap/kvproto/pull/737)
- [x] merge [kvproto GRPC service definitions upstream](https://github.com/pingcap/kvproto/pull/748)

WIP:
- [X] test external_storage against S3
- [X] test external_storage against GCS
- [X] test kvproto change

### What problem does this PR solve?

We are organizing the cloud code in a separate crate. This will facilitate the cloud team being able to own the code and modifying builds to include or exclude cloud providers or in the future dynamically link them.

The encryption code reoganization is already merged here: https://github.com/tikv/tikv/pull/9647
This PR is for the external_storage code.


### What is changed and how it works?


* AWS and GCP code are moved from the external_storage crate into the cloud/aws and cloud/gcp crates
* components/external_storage/export is added to avoid cyclical dependencies. Cloud constructors of external_storage traits are imported here and then re-exported along with the external_storage crate for end users of external_storage
* rusoto_util code was already moved to aws/util in the encryption PR and in this PR we just delete the rusoto_util crate
* There is a flag `cloud-storage-grpc` for an experimental mode where external storage is served from a separate GRPC service over a unix domain socket.
* There is a flag `cloud-storage-dylib` for an experimental mode where external storage is called via C bindings. This code uses most of the same code from the GRPC flag. This code can be linked statically or loaded dynamically.


### Check List

Tests

- Unit test
- Manual test (WIP)

scli with localstack as s3. Tested both a text file (README) and a binary file (scli)

```
AWS_ACCESS_KEY_ID=test AWS_SECRET_ACCESS_KEY=test ./target/debug/examples/scli --storage S3 --endpoint http://localhost:4566 --region us-west-2 --bucket backup --name ext_storage --path . --file README.md save
aws s3 --endpoint http://localhost:4566 cp s3://backup/ext_storage .
diff ext_storage README.md
```

br with localstack as S3. Test dump a database, delete the database, then restore it. Restore uses same arguments as backup. Also test this after building with the flag `cloud-storage-dylib`.

```
AWS_ACCESS_KEY_ID=test AWS_SECRET_ACCESS_KEY=test ~/go/src/github.com/pingcap/br/bin/br --version backup full -s s3://backup/full2 --s3.endpoint http://localhost:4566 --s3.provider aws --s3.region us-west-2
```

br with fake-gcs server as gcs (uses some flags that I recently added on master).

```
br --gcs.endpoint http://storage.gcs.127.0.0.1.nip.io:4443/ --no-credentials --send-credentials-to-tikv=false --skip-check-path backup -s gcs://backup/alpha full
```

Delete the database from TiDB, after restore see that it exists again

```
br --gcs.endpoint http://storage.gcs.127.0.0.1.nip.io:4443/ --no-credentials --send-credentials-to-tikv=false restore -s gcs://backup/alpha full
```

Test cloud dynamic protobuf. Tested for AWS S3 as well. Tested setting both --host and --endpoint for GCS.
Below is the save command. Then test by doing the load command and performing a diff.

```
./target/debug/examples/scli --storage cloud --cloud-name gcp --host http://localhost:4443/ --bucket backup --name ext_storage9 --path . --file README.md save
```

Test experimental GRPC service. Run backup with binaries running that are built from these steps:

```
ENABLE_FEATURES=cloud-storage-grpc make build
./scripts/env cargo build -p external_storage_export --features=protobuf-codec,cloud-gcp,cloud-aws,cloud-storage-grpc && RUST_BACKTRACE=1 RUST_LIB_BACKTRACE=1 ./target/debug/tikv-cloud-storage
```

Dynamically loading the external storage code:

```
# manually edit the Cargo manifest to build a cdylib

# now build the lib
./scripts/env cargo build -p external_storage_export --features=cloud-storage-dylib,cloud-gcp,cloud-aws,protobuf-codec

# ensure the lib is int the working directory of tikv-server
cp target/debug/libexternal_storage_export.so .
```

If the lib cannot be dynamically loaded, you will see warnings during the backup in the logs.

Compatibilty
* Programs built from the TiKV lib (tikv-server, etc) now import the encryption_export crate instead of encryption directly.


### Release note

Move external_storage cloud provider code out of external_storage and into separate cloud provider crates.
